text/0000-optimise-attr.md

```diff
@@ -0,0 +1,120 @@
+
+The `#[optimise(size)]` attribute applied to a function definition will instruct the optimisation
+engine to avoid applying optimisations that could result in a size increase and machine code
+generator to generate code that’s smaller rather than larger.
+
+Note that the `#[optimise(size)]` attribute is just a hint and is not guaranteed to result in any
+different or smaller code.
+
+Since `#[optimise(size)]` instructs optimisations to behave in a certain way, this means that this
+attribute has no effect when no optimisations are run (such as is the case when `-Copt-level=0`).
+Interaction of this attribute with the `-Copt-level=s` and `-Copt-level=z` flags is not specified
+and is left up to implementation to decide.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* Not all of the alternative codegen backends may be able to express such a request, hence the
+“this is an optimisation hint” note on the `#[optimise(size)]` attribute.
+    * As a fallback, this attribute may be implemented in terms of more specific optimisation hints
+      (such as `inline(never)`, the future `unroll(never)` etc).
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+Proposed is a very semantic solution (describes the desired result, instead of behaviour) to the
+problem of needing to sometimes inhibit some of the trade-off optimisations such as loop unrolling.
+
+Alternative, of course, would be to add attributes controlling such optimisations, such as
+`#[unroll(no)]` on top of a a loop statement. There’s already precedent for this in the `#[inline]`
+annotations.
+
+The author would like to argue that we should eventually have *both*, the `#[optimise(size)]` for
+people who look at generated code and decide that it is too large, and the targetted attributes for
+people who know *why* the code is too large.
+
+Furthermore, currently `optimise(size)` is able to do more than any possible combination of
+targetted attributes would be able to such as influencing the instruction selection or switch
+codegen strategy (jump table, if chain, etc.) This makes the attribute useful even in presence of
+all the targetted optimisation knobs we might have in the future.
+
+---
+
+Alternative: `optimize` (American English) instead of `optimise`… or both?
+
+# Prior art
+[prior-art]: #prior-art
+
+* LLVM: `optsize`, `optnone`, `minsize` function attributes (exposed in Clang in some way);
+* GCC: `__attribute__((optimize))` function attribute which allows setting the optimisation level
+and using certain(?) `-f` flags for each function;
+* IAR: Optimisations have a checkbox for “No size constraints”, which allows compiler to go out of
+its way to optimise without considering the size tradeoff. Can only be applied on a
+per-compilation-unit basis. Enabled by default, as is appropriate for a compiler targetting
+embedded use-cases.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* Should we support such an attribute at module-level? Crate-level?
```

Comment:
I think so yes. At least, you should be able to specify `#[optimize(size)]` on `mod` and `impl` with the semantics that it applies to every `fn` inside those transitively.

Reply:
Counter point: people might assume that `optimise(size)` at the crate level also asserts `repr(packed)`. We don't do this with `inline` so I'd assume not for optimise.

Reply:
I think it would be quite tedious to have to do this on every single function if a module if that is what you want. If we want to be more clear that it is about functions, you could write `#[optimize(fn_size)] mod foobar { .. }`.

Reply:
I do think it makes sense to allow `#![optimize(size)]` in a crate to apply to all functions (that don't override it).

I *certainly* don't think `optimize(size)` should ever imply `repr(packed)`, because the latter can affect semantics.